generator/restrictions.hpp

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include "generator/osm_id.hpp"
+
+#include "std/functional.hpp"
+#include "std/limits.hpp"
+#include "std/mutex.hpp"
+#include "std/unordered_map.hpp"
+#include "std/utility.hpp"
+#include "std/vector.hpp"
+
+class RelationElement;
+
+template <> struct hash<osm::Id>
+{
+  size_t operator()(osm::Id const & id) const
+  {
+    return hash<uint64_t>()(id.OsmId());
+  }
+};
+
+/// This class collects all relations with type restriction and save feature ids of
+/// their road feature in text file for using later.
+class RestrictionCollector
+{
+  friend void UnitTest_RestrictionTest_ValidCase();
+  friend void UnitTest_RestrictionTest_InvalidCase();
+public:
+  using FeatureId = uint32_t;
+  static FeatureId const kInvalidFeatureId;
+
+  /// \brief Types of road graph restrictions.
+  /// \note Despite the fact more that 10 restriction tags are present in osm all of them
+  /// could be split into two categories.
+  /// * no_left_turn, no_right_turn, no_u_turn and so on go to "No" category.
+  /// * only_left_turn, only_right_turn and so on go to "Only" category.
+  /// That's enough to rememeber if
+  /// * there's only way to pass the junction is driving along the restriction (Only)
+  /// * driving along the restriction is prohibited (No)
+  enum class Type
+  {
+    No, // Going according such restriction is prohibited.
+    Only, // Only going according such restriction is permitted
+  };
+
+  /// \brief Restriction to modify road graph.
+  struct Restriction
+  {
+    Restriction(Type type, size_t linkNumber);
+    // Constructor for testing.
+    Restriction(Type type, vector<FeatureId> const & links);
```

Comment:
Now the situation is changed. It's used for initializing Restriction after Deserializtion. The comment was removed.